stan/math/prim/mat/functor/mpi_parallel_call.hpp

```diff
@@ -0,0 +1,488 @@
+    // local jobs to the root. This needs to be done for the number of
+    // outputs of each result and the number of outputs per job.
+    if (num_outputs_per_job_ == -1) {
+      // before we can cache the sizes locally we must ensure
+      // that no exception has been fired from any node. Hence,
+      // share the info about the current status across all nodes.
+      int cluster_status = 0;
+      boost::mpi::reduce(world_, local_ok, cluster_status, std::plus<int>(), 0);
+      bool all_ok = cluster_status == static_cast<int>(world_size_);
+      boost::mpi::broadcast(world_, all_ok, 0);
+      if (!all_ok) {
+        // err out on the root
+        if (rank_ == 0) {
+          throw std::domain_error("MPI error on first evaluation.");
+        } else {
+          // and ensure on the workers that they return into their
+          // listening state
+          return result_t();
+        }
+      }
+
+      // execution was ok everywhere such that we can now distribute the
+      // outputs per job as recorded on the root on the first execution
+      boost::mpi::broadcast(world_, local_outputs_per_job, 0);
+      num_outputs_per_job_ = local_outputs_per_job;
+
+      if (!cache_f_out::is_valid()) {
+        // ok, now all is fine everywhere, collect sizes and cache info
+        std::vector<int> world_f_out(num_jobs, 0);
+        boost::mpi::gatherv(world_, local_f_out.data(), num_local_jobs,
+                            world_f_out.data(), job_chunks, 0);
+        // on the root we now have all sizes from all childs. Copy
+        // over the local sizes to the world vector on each local
+        // node in order to cache this information locally.
+        std::copy(local_f_out.begin(), local_f_out.end(),
+                  world_f_out.begin() + start_job);
+        cache_f_out::store(world_f_out);
+      }
+    }
+
+    typename cache_f_out::cache_t& world_f_out = cache_f_out::data();
+
+    // check that cached sizes are the same as just collected from
+    // this evaluation
+    for (int i = 0; i < num_local_jobs; ++i) {
+      if (world_f_out[start_job + i] != local_f_out[i]) {
+        local_ok = 0;
+        break;
+      }
+    }
+
+    const std::size_t size_world_f_out = sum(world_f_out);
+    matrix_d world_result(num_outputs_per_job_, size_world_f_out);
+
+    std::vector<int> chunks_result(world_size_, 0);
+    for (std::size_t i = 0, ij = 0; i != world_size_; ++i)
+      for (int j = 0; j != job_chunks[i]; ++j, ++ij)
+        chunks_result[i] += world_f_out[ij] * num_outputs_per_job_;
+
+    // collect results on root
+    boost::mpi::gatherv(world_, local_output.data(), chunks_result[rank_],
+                        world_result.data(), chunks_result, 0);
+
+    // let root know if all went fine everywhere
+    int cluster_status = 0;
+    boost::mpi::reduce(world_, local_ok, cluster_status, std::plus<int>(), 0);
+    bool all_ok = cluster_status == static_cast<int>(world_size_);
```

Comment:
I think it `local_ok` being bad will cause `all_ok` to be bad?  If so, I don't know why it's being called here.  If this ust sets `local_ok`, then it doesn't get used anywhere and probably should.

Reply:
Here I am adding up the local_ok from all workers. If that sum is equal to the world_size, then (and only then) all workers had local_ok==1.